[Test] 테스트 구현

dev-msj opened the pull request:
## 작업 내용
controller/home/HomeController.java 구조에서 controller/HomeController.java 구조로 패키지 단순화
Mapper 파일들이 Table 단위로 사용하도록 변경
BeforeEach, AfterEach를 통해 test 데이터를 사용하도록 변경
EhCache와 Redis 모두 CacheManager로 등록되도록 변경
BookSuggestionRepository 로직 최적화
HomeController에서 Service에 suggestion을 요청 시 BookSuggestionRequest Model을 활용하도록 변경
HomeController 테스트 추가

## 이슈 번호
#17

dev-msj committed:
style : 패키지 경로 변경

dev-msj committed:
refactor : user 관련 내용은 UserMapper에서 처리되도록 변경

dev-msj committed:
fix : UserMapper namespace 경로 에러 수정

dev-msj committed:
feat : UserAuth, UserInfo, UserCategory 삭제 기능 추가

dev-msj committed:
test : BeforeEach, AfterEach를 통해 mock user 데이터를 사용하도록 변경

dev-msj committed:
refactor : BookSuggestionRepository 로직 최적화

suggestionType 별로 redis에 캐싱하도록 변경
HomeMapper에서 suggestionType별로 select하도록 변경

dev-msj committed:
refactor : EhCache와 Redis 모두 CacheManager로 등록되도록 변경

org.ehcache에서 net.sf.ehcache로 gradle 변경
ehcache.xml을 net.sf.ehcache에 해당하는 형식으로 변경
RedisConfig에 redisCacheManager Bean 추가
EhCacheConfig에 EhCacheCacheManager Bean 추가
application.yml에서 cache 설정 삭제

dev-msj committed:
refactor : HomeController에서 Service에 suggestion을 요청 시 Model을 활용하도록 변경

OasisKeyGenerator에서 Model을 받아 key를 생성하도록 변경

dev-msj committed:
test : HomeController 테스트 추가

dev-msj committed:
refactor : Mapper를 Table 단위로 사용하도록 변경

jinspark-lab commented on src/main/java/com/flab/oasis/config/OasisKeyGenerator.java:
```diff
@@ -15,7 +16,10 @@ public Object generate(Object target, Method method, Object... params) {
+            if (obj instanceof BookSuggestionRequest) {
+                stringList.add(((BookSuggestionRequest) obj).getUid());
+                stringList.add(((BookSuggestionRequest) obj).getSuggestionType().name());
+            }
```
InstanceType 마다 매번 조건문을 구현하실건가요? 추상화 아이디어를 생각해보시면 좋을 것 같습니다.

jinspark-lab replied:
BaseRequest 로 묶는 것도 좋지만 이런 용도의 인터페이스를 따로 만들어두고, 해당 인터페이스의 generateEhCacheKey() 를 그냥 사용하게 하면 조건문 자체를 아예 없앨 수 있습니다.

jinspark-lab commented on src/main/java/com/flab/oasis/model/UserAuth.java:
```diff
@@ -0,0 +1,16 @@
+    private String password;
+    private char socialYN;
+    private String accessToken;
+    private String refreshToken;
```
Test 에 상관없는건 빼셔야합니다. 로그인은 아직 구현 안된 Feature 니까요.. 더미로 구현하시고 클래스화는 빼시는게 맞습니다.

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -1,46 +1,100 @@
+        userAuthMapper.insertUserAuth(
+                new UserAuth(uid, "1234", 'N', null, null)
+        );
+        userInfoMapper.insertUserInfo(
```
구현 안된 기능을 테스트 때문에 끼워넣으시는건 좋지않습니다. 차라리 하드코딩으로 uid 를 더미 데이터로 넣는게 낫습니다.

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -1,46 +1,100 @@
+                MockMvcRequestBuilders.post("/home/suggestion")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(bookSuggestionRequest))
+        ).andExpect(MockMvcResultMatchers.status().isOk()).andExpect(MockMvcResultMatchers.content().json(expected));
```
API 를 호출한 결과 JSON 이랑 Service 에서 반환된 결과값을 테스트하시는 건가요?

jinspark-lab reviewed (changes requested):
로그인 관련 기능 전부 빼주시고, 더미데이터로 테스트해주세요. 로그인 모델이 Home Test 에 들어가시면 나중에 추적이 안되십니다.
테스트코드는 작성하는 연습을 하셔야할 것 같습니다.. 이전에 몇번 지적했던 부분인데 의미없는 테스트 코드를 계속 작성하고 계세요. 
관련해서는 수업시간에 좀 더 보강드리겠습니다.

dev-msj committed:
refactor : OasisKeyGenrator의 키 생성 로직 추상화

캐싱이 필요한 Request는 BaseRequest를 상속받아 cache key 생성 로직을 구현하도록 변경

dev-msj committed:
test : Mockito를 적용하여 테스트하도록 변경

jinspark-lab commented on src/main/java/com/flab/oasis/model/BaseRequest.java:
```diff
@@ -0,0 +1,5 @@
+package com.flab.oasis.model;
+
+public interface BaseRequest {
+    public String combineKeyWithUnderBar();
```
generateRedisKey 같은 네이밍이 좋아보입니다

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -1,46 +1,118 @@
+//                        .get(String.format("%s_%s_%s", "suggestion", uid, suggestionType));
+//
+//        Assertions.assertEquals(expected, Objects.requireNonNull(actual).get());
+//    }
```
이런 불필요한 주석은 지워주세요

jinspark-lab commented on src/main/java/com/flab/oasis/repository/BookSuggestionRepository.java:
```diff
@@ -0,0 +1,39 @@
+    public List<BookSuggestion> getBookSuggestionList(SuggestionType suggestionType) {
+        Object stringValue = redisTemplate.opsForValue().get(suggestionType);
+
+        if (Objects.isNull(stringValue)) {
```
Optional.ofNullable 이 더 깔끔해보입니다.

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeControllerTest.java:
```diff
@@ -0,0 +1,97 @@
+        bookSuggestionRequest.setUid(uid);
+        bookSuggestionRequest.setSuggestionType(suggestionType);
+
+        return bookSuggestionRequest;
```
재사용 가능성이 없는데 굳이 메서드로 안빼셔도 될 것 같습니다. 차라리 AllArgsConstructor 를 이용하는게 좋아보입니다

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeControllerTest.java:
```diff
@@ -0,0 +1,97 @@
+            bookSuggestion.setDescription(String.format("%s%d", "desc", i));
+            bookSuggestion.setImageUrl(String.format("%s%d", "url", i));
+
+            bookSuggestionList.add(bookSuggestion);
```
굳이 5개나 만드시는 이유가 있나요?

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeControllerTest.java:
```diff
@@ -0,0 +1,97 @@
+        resultActions.andExpect(MockMvcResultMatchers.status().isOk())
+                .andExpect(MockMvcResultMatchers.content().json(objectMapper.writeValueAsString(bookSuggestionList)));
+
+    }
```
리뷰 커멘트로 대신하겠습니다.

jinspark-lab reviewed (changes requested):
이전에 참고하시라고 소스 리포지토리들을 공유드렸는데 안보신 것 같습니다. 참고하시고 테스트쪽 문제점 파악해보시고 수정해주세요.

dev-msj committed:
test : user category가 존재하지 않는 케이스 추가

dev-msj committed:
style : BookSuggestionRequest에 AllArgsConstructor 추가

dev-msj committed:
test : BookSuggestionRequest가 생성자를 통해 초기화도록 변경

dev-msj committed:
style : BaseRequest의 메서드명을 generateEhCacheKey로 변경

dev-msj committed:
refactor : BookSuggestionRepository에서 Objects.isNull을 Optional.ofNullable로 변경

dev-msj committed:
fix : method명 미스매치 오류 수정

dev-msj committed:
test : 메서드명에 테스트 의도를 설명하도록 변경

dev-msj committed:
style : 사용되지 않는 Exception 삭제

dev-msj committed:
test(HomeControllerTest) : 결과 비교를 특정 index의 특정값만 비교하도록 변경

dev-msj committed:
style(HomeServiceTest) : 메서드명 수정

dev-msj committed:
fix : 현재 feature와 상관없는 코드 삭제

dev-msj committed:
fix(ConnectionTest) : redis 테스트 오류 수정

dev-msj committed:
style(CommonTest) : ConnectionTest를 CommonTest로 변경

dev-msj committed:
style(HomeService) : UserCategoryMapper의 변수명을 클래스명과 동일하게 변경

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeControllerTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @DisplayName("/home/suggestion POST 요청")
+    @Test
+    void response된_값과_실제값_비교() throws Exception {
```
이건 뭔가요?

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -1,46 +1,104 @@
-    void ehCacheTest() {
-        List<BookSuggestion> bookList =  homeService.suggestion(
-                "test@naver.com", SuggestionType.valueOf("newBook".toUpperCase())
+    void 유저_카테고리에_해당하는_카테고리만_반환() {
```
한글로 입력된 부분 일부러 이렇게 입력하신건가요?

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -1,46 +1,104 @@
+            bookSuggestion.setDescription(String.format("%s%d", "desc", i));
+            bookSuggestion.setImageUrl(String.format("%s%d", "url", i));
+
+            bookSuggestionList.add(bookSuggestion);
```
더미값을 굳이 루프를 돌리는 이유를 모르겠습니다.

jinspark-lab commented on src/test/java/com/flab/oasis/home/HomeServiceTest.java:
```diff
@@ -1,46 +1,104 @@
-        Cache cache = cacheManager.getCache("homeCache");
+    @DisplayName("User Category가 존재하지 않을 때")
+    @Test
+    void 유저_카테고리가_없을_때_모든_카테고리를_반환() {
```
마찬가지로 한글 테스트명은 왜 쓰시는건가요

jinspark-lab reviewed (commented):
코드 자체는 피드백 받고 나아졌는데.. 스타일에 대해서 좀 점검이 필요해보입니다. 수업시간에 피드백드릴게요.

dev-msj committed:
test(HomeServiceTest) : 더미 값 루프 제거

dev-msj committed:
test(HomeServiceTest) : 각 테스트 의도에 맞는 검증을 진행하도록 변경

User Category가 존재할 때
- 제공된 UserCategory의 categoryId가 실제 반환된 데이터에 존재하는지 확인
User Category가 존재하지 않을 때
- 추천 책들 중 0번째 있는 categoryId가 실제 반환된 데이터에 존재하는지 확인

dev-msj committed:
style : 테스트 코드 메서드명을 영어로 변경
